PR: https://github.com/odoo/odoo/pull/241298

From: 6c145db4b07b7f7ab1dbcea3f7257a518cfc7dcb
From: tpma-odoo
Date: 2026-06-16 06:25:11

Breaking data model changes scores: del:1 + add:1, change matches:
-    lot_id = fields.Many2one(
+    lot_ids = fields.Many2many(

Total Changes: 62

[IMP] mrp: allow unbuilding multiple serials in a single unbuild order

* Currently, the unbuild order only allows a single serial number to be unbuilt,
although manufacturing orders can now produce multiple serial numbers. This
limitation forces users to create separate unbuild orders for each serial, which
is time-consuming and leads to a poor user experience for serial-tracked product

* This commit allows users to unbuild multiple serial numbers in a single
unbuild order. Users can now select and process multiple serials together,
allowing them to reverse manufacturing orders more efficiently without creating
multiple unbuild orders. The existing behavior for lot-tracked products is
preserved by allowing only one lot per unbuild order, with a clear action to
reset the selected lot or serials. In addition, the dedicated simplified unbuild
view is removed in favor of the standard unbuild form, reducing maintenance
and avoiding the need to maintain separate unbuild flows. This improves
efficiency, simplifies the workflow and enhances usability when working with
serial-tracked products.

Upgrade PR: odoo/upgrade#9145
TaskID-5350021

closes odoo/odoo#241298

Signed-off-by: Dhaval Hadiya (dhha) <dhha@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -2579,19 +2579,19 @@ class MrpProduction(models.Model):
     def button_unbuild(self):
         self.ensure_one()
         return {
-            'name': _('Unbuild: %s', self.product_id.display_name),
             'view_mode': 'form',
             'res_model': 'mrp.unbuild',
-            'view_id': self.env.ref('mrp.mrp_unbuild_form_view_simplified').id,
             'type': 'ir.actions.act_window',
-            'context': {'default_product_id': self.product_id.id,
-                        'default_lot_id': self.lot_producing_ids[:1].id,
-                        'default_mo_id': self.id,
-                        'default_company_id': self.company_id.id,
-                        'default_location_id': self.location_dest_id.id,
-                        'default_location_dest_id': self.location_src_id.id,
-                        'create': False, 'edit': False},
-            'target': 'new',
+            'context': {
+                'default_product_id': self.product_id.id,
+                'default_product_qty': self.qty_producing,
+                'default_lot_ids': self.lot_producing_ids.ids,
+                'default_mo_id': self.id,
+                'default_company_id': self.company_id.id,
+                'default_location_id': self.location_dest_id.id,
+                'default_location_dest_id': self.location_src_id.id,
+                'create': False,
+            },
         }
 
     def action_split(self):

--- a/addons/mrp/models/mrp_unbuild.py
+++ b/addons/mrp/models/mrp_unbuild.py
@@ -2,7 +2,7 @@
 
 from collections import defaultdict
 
-from odoo import _, api, fields, models
+from odoo import Command, _, api, fields, models
 from odoo.exceptions import UserError
 from odoo.tools import float_compare, float_round
 from odoo.tools.misc import clean_context
@@ -27,7 +27,7 @@ class MrpUnbuild(models.Model):
     product_qty = fields.Float(
         'Quantity', default=1.0,
         digits='Product Unit',
-        compute='_compute_product_qty', store=True, precompute=True, readonly=False,
+        compute='_compute_product_qty', store=True, readonly=False,
         required=True)
     uom_id = fields.Many2one(
         'uom.uom', 'Unit',
@@ -54,8 +54,8 @@ class MrpUnbuild(models.Model):
         check_company=True, index='btree_not_null')
     mo_bom_id = fields.Many2one('mrp.bom', 'Bill of Material used on the Production Order', related='mo_id.bom_id')
     lot_producing_ids = fields.Many2many('stock.lot', string='Lot/Serial Numbers', related='mo_id.lot_producing_ids')
-    lot_id = fields.Many2one(
-        'stock.lot', 'Lot/Serial Number',
+    lot_ids = fields.Many2many(
+        'stock.lot', string='Lot/Serial Number',
         domain="[('product_id', '=', product_id),('id', 'in', lot_producing_ids)]", check_company=True)
     has_tracking = fields.Selection(related='product_id.tracking', readonly=True)
     location_id = fields.Many2one(
@@ -123,14 +123,13 @@ class MrpUnbuild(models.Model):
             if order.mo_id and order.mo_id.product_id:
                 order.product_id = order.mo_id.product_id
 
-    @api.depends('mo_id')
+    @api.depends('mo_id', 'lot_ids')
     def _compute_product_qty(self):
         for order in self:
-            if order.mo_id:
-                if order.has_tracking == 'serial':
-                    order.product_qty = 1
-                else:
-                    order.product_qty = order.mo_id.qty_produced
+            if order.has_tracking == 'serial':
+                order.product_qty = len(order.lot_ids) or 1
+            elif order.mo_id:
+                order.product_qty = order.mo_id.qty_produced
 
     @api.model_create_multi
     def create(self, vals_list):
@@ -145,15 +144,17 @@ class MrpUnbuild(models.Model):
             raise UserError(_("You cannot delete an unbuild order if the state is 'Done'."))
 
     def _prepare_finished_move_line_vals(self, finished_move):
-        return {
+        vals = {
             'move_id': finished_move.id,
-            'lot_id': self.lot_id.id,
-            'quantity': finished_move.product_uom_qty - finished_move.quantity,
             'product_id': finished_move.product_id.id,
             'uom_id': finished_move.uom_id.id,
             'location_id': finished_move.location_id.id,
             'location_dest_id': finished_move.location_dest_id.id,
         }
+        if len(self.lot_ids) > 1:
+            return [dict(vals, lot_id=lot.id, quantity=1) for lot in self.lot_ids]
+        else:
+            return [dict(vals, lot_id=self.lot_ids.id, quantity=finished_move.product_uom_qty - finished_move.quantity)]
 
     def _prepare_move_line_vals(self, move, origin_move_line, taken_quantity):
         return {
@@ -166,12 +167,15 @@ class MrpUnbuild(models.Model):
             'location_dest_id': move.location_dest_id.id,
         }
 
+    def action_clear_lots(self):
+        self.lot_ids = [Command.clear()]
+
     def action_unbuild(self):
         self.ensure_one()
         self._check_company()
         # remove the default_* keys that were only needed in the unbuild wizard
         self = self.with_env(self.env(context=clean_context(self.env.context)))  # noqa: PLW0642
-        if self.product_id.tracking in ['lot', 'serial'] and not self.lot_id.id:
+        if self.product_id.tracking in ['lot', 'serial'] and not self.lot_ids:
             raise UserError(_('You should provide a lot number for the final product.'))
 
         if self.mo_id and self.mo_id.state != 'done':
@@ -216,9 +220,9 @@ class MrpUnbuild(models.Model):
                 continue
             needed_quantity = move.product_uom_qty
             moves_lines = original_move.mapped('move_line_ids')
-            if move in produce_moves and self.lot_id:
+            if move in produce_moves and self.lot_ids:
                 moves_lines = moves_lines.filtered(
-                    lambda ml: self.lot_id in ml.produce_line_ids.lot_id and ml.lot_id not in previously_unbuilt_lots
+                    lambda ml: ml.produce_line_ids.lot_id & self.lot_ids and ml.lot_id not in previously_unbuilt_lots
                 )
             for move_line in moves_lines:
                 # Iterate over all move_lines until we unbuilded the correct quantity.
@@ -324,8 +328,12 @@ class MrpUnbuild(models.Model):
     def action_validate(self):
         self.ensure_one()
         precision = self.env['decimal.precision'].precision_get('Product Unit')
-        available_qty = self.env['stock.quant']._get_available_quantity(self.product_id, self.location_id, self.lot_id, strict=True)
+        available_qty = 0
         unbuild_qty = self.uom_id._compute_quantity(self.product_qty, self.product_id.uom_id)
+        if len(self.lot_ids) > 1:
+            available_qty = sum(self.env['stock.quant']._get_available_quantity(self.product_id, self.location_id, lot, strict=True) for lot in self.lot_ids)
+        else:
+            available_qty = self.env['stock.quant']._get_available_quantity(self.product_id, self.location_id, self.lot_ids, strict=True)
         if float_compare(available_qty, unbuild_qty, precision_digits=precision) >= 0:
             return self.action_unbuild()
         else:
